Changes to allow enemies to climb

syntaxi opened the pull request:
Utilises a plugin to FlexiblePathfinding to allow enemies to travel vertically. The enemies can only travel up or down one block at a time.
Also includes a fix for an odd bug where calculating the path by starting at the shrine caused it to not correctly function.

## Testing

Test by simply loading any new world and activating it by interacting with any block (via `E`). The displayed path will travel over one block high obstacles.

Placing a two block high wall in the way will cause the path to go around it (Or over if it can find a one high block to use to get ontop of the wall)

Commit:
Add a plugin that allows gooey to go up and down one block height differences

Commit:
Display path as a two block high transparent path

Commit:
Tweak enemy walk path to not only allow diagonal movement in the horizontal axis

Commit:
Fix bizzare bug where requesting the path backwards would cause it to not correctly calculate

oniatus commented on src/main/java/org/terasology/gooeyDefence/PathfindingSystem.java:
```diff
@@ -125,7 +130,18 @@ private void calculatePath(int id, Runnable callback) {
+        pathfinderSystem.requestPath(config, (path, end) -> {
+            Collections.reverse(path);
+            callback.accept(path, DefenceField.fieldCentre());
+        });
```
-you can separate the config creation from the path request and callback in two methods.
-extract magic numbers to constants with speaking names, e.g. what is 10f or 0.5f
-separate defenceField access from calculation, pass these values as parameters.

oniatus commented on src/main/java/org/terasology/gooeyDefence/pathfinding/EnemyWalkingPlugin.java:
```diff
@@ -0,0 +1,58 @@
+                /* Check that at least one of the to or from are walkable */
+                || !(isWalkable(to) || isWalkable(from))) {
+            return false;
+        }
```
`if(areEqual(from,to) || !isDiagonalReachable(from,to) || !isStraightReachable(from,to) || !isWalkable(from,to)` could be more readable.

`if(x) return false` can be refactored to `return !x`. It could make sense to refactor the method without negation, e.g isReachable could be `notEqual(from,to) && reachable(from,to) && walkable(from,to)`.

The walkable condition looks a bit odd.
In this context it would be
isReachable == false if !(isWalkable(to) || isWalkable(from)).
-> isReachable = isWalkable(to) || isWalkable(from). This looks like it should be an && instead.

niksoc-hi replied:
the condition does look odd. But i think that it's used for falling off ledges and jumping. In any case it can be clarified. And again if possible it would be great to refactor flexible path finding too (at least adding explanatory comments) where I see the same condition.

niksoc replied:
the condition does look odd. But i think that it's used for falling off ledges and jumping. In any case it can be clarified. And again if possible it would be great to refactor flexible path finding too (at least adding explanatory comments) where I see the same condition.

oniatus commented on src/main/java/org/terasology/gooeyDefence/pathfinding/EnemyWalkingPlugin.java:
```diff
@@ -0,0 +1,58 @@
+            return false;
+        }
+
+        /* Check that all blocks passed through by this movement are penetrable */
```
`return areAllPassedThroughBlocksPenetrable(...)` ;)

niksoc-hi replied:
aren't these the exact same lines as in the overridden superclass method? Is it possible to refactor the flexiblepathfinding code (extract these lines like oniatus says) and reuse the code from the superclass?

niksoc replied:
aren't these the exact same lines as in the overridden superclass method? Is it possible to refactor the flexiblepathfinding code (extract these lines like oniatus says) and reuse the code from the superclass?

oniatus commented on src/main/java/org/terasology/gooeyDefence/pathfinding/EnemyWalkingPlugin.java:
```diff
@@ -0,0 +1,58 @@
+        }
+
+        /* Check that all blocks passed through by this movement are penetrable */
+        for (Vector3i occupiedBlock : getOccupiedRegionRelative()) {
```
strange name, is this a block position or a relative position? In both cases it is no real "block" ;)

oniatus commented on src/main/java/org/terasology/gooeyDefence/pathfinding/EnemyWalkingPlugin.java:
```diff
@@ -0,0 +1,58 @@
+            Vector3i occupiedBlockFrom = new Vector3i(from).add(occupiedBlock);
+
+            Region3i movementBounds = Region3i.createBounded(occupiedBlockTo, occupiedBlockFrom);
+            for (Vector3i block : movementBounds) {
```
same here =)

niksoc-hi commented on src/main/java/org/terasology/gooeyDefence/PathfindingSystem.java:
```diff
@@ -125,7 +130,18 @@ private void calculatePath(int id, Runnable callback) {
     private void calculatePath(Vector3i startPoint, BiConsumer<List<Vector3i>, Vector3i> callback) {
-        pathfinderSystem.requestPath(DefenceField.fieldCentre(), startPoint, callback::accept);
+        JPSConfig config = new JPSConfig();
+        /* Backwards to ensure that zero is at the field centre */
```
can't understand this comment. What backwards?

niksoc replied:
can't understand this comment. What backwards?

Commit:
Refactor checks in `EnemyWalkingPlugin` into seperate methods.

Commit:
Refactor pathfinding config creation into a seperate method.
Change callback used from `BiConsumer` to `Consumer` as the second parameter was never used.

syntaxi commented:
Changes have been made. I shall now merge this PR ahead of the coming content additions.
